styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java

```diff
@@ -102,64 +109,67 @@ public WorkflowResource(Storage storage, String schedulerServiceBaseUrl,
-                                                                       RequestContext rc) {
-    // TODO: handle validation in one place, see SchedulerResource.java
+  private Response<Workflow> createOrUpdateWorkflow(String componentId,
+                                                    RequestContext rc) {
     final Optional<ByteString> payload = rc.request().payload();
     if (!payload.isPresent()) {
-      return CompletableFuture.completedFuture(
-          Response.forStatus(Status.BAD_REQUEST.withReasonPhrase("Missing payload.")));
+      return Response.forStatus(BAD_REQUEST.withReasonPhrase("Missing payload."));
     }
     final WorkflowConfiguration workflowConfig;
     try {
       workflowConfig = OBJECT_MAPPER
           .readValue(payload.get().toByteArray(), WorkflowConfiguration.class);
     } catch (IOException e) {
-      return CompletableFuture.completedFuture(
-          Response.forStatus(Status.BAD_REQUEST
-                                 .withReasonPhrase("Invalid payload. " + e.getMessage())));
+      return Response.forStatus(BAD_REQUEST
+          .withReasonPhrase("Invalid payload. " + e.getMessage()));
     }
 
-    final Collection<String> errors = workflowValidator.validateWorkflowConfiguration(workflowConfig);
+    final Collection<String> errors =
+        workflowValidator.validateWorkflowConfiguration(workflowConfig);
     if (!errors.isEmpty()) {
-      return CompletableFuture.completedFuture(
-          Response.forStatus(Status.BAD_REQUEST.withReasonPhrase("Invalid workflow configuration: " + errors)));
+      return Response.forStatus(
+          BAD_REQUEST.withReasonPhrase("Invalid workflow configuration: " + errors));
+    }
+
+    final Workflow workflow = Workflow.create(componentId, workflowConfig);
+    final Optional<Workflow> oldWorkflowOptional;
+    try {
+      oldWorkflowOptional = workflowInitializer.inspectChange(workflow);
+    } catch (WorkflowInitializationException e) {
+      return Response.forStatus(BAD_REQUEST.withReasonPhrase(e.getMessage()));
     }
 
-    // TODO: handle workflow crud directly in api service instead of proxying to scheduler
-    return forwardingClient.send(rc.request()
-                                     .withPayload(payload.get())
-                                     .withUri(schedulerApiUrl("workflows", componentId)));
+    if (oldWorkflowOptional.isPresent()) {
+      LOG.info("Workflow modified, old config: {}, new config: {}", oldWorkflowOptional.get(),
+          workflow);
+    } else {
+      LOG.info("Workflow added: {}", workflow);
+    }
+
+    return Response.forPayload(workflow);
   }
 
   private Response<Collection<Workflow>> workflows() {
     try {
       return Response.forPayload(storage.workflows().values());
     } catch (IOException e) {
-      return Response.forStatus(Status.INTERNAL_SERVER_ERROR.withReasonPhrase(
+      return Response.forStatus(INTERNAL_SERVER_ERROR.withReasonPhrase(
```

Comment:
Would've been great to have this noise due to static import change in a different PR

Reply:
Sure will get them back.